### PR TITLE
Api complete with database

### PR DESCRIPTION
**Pull request by Rwothoromo**
The Api works with a database as per the url on the readme and on postman.

Requesting final reviews @collin5

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -1,10 +1,72 @@
...
+from config import app_config
+from app.db import db
+
+from app.api.v2.resources.hello import HelloWorld
+from app.api.v2.resources.auth import RegisterUser, LoginUser, ResetPassword, LogoutUser
+from app.api.v2.resources.business import BusinessCollection, BusinessResource, BusinessReviews
 
-# Initialize the app
 app = Flask(__name__)
+app.config.from_object(app_config[os.environ['FLASK_CONFIG']])
```
Comment (collin5) [inline app/__init__.py]: Use `os.environ.get('FLASK_CONFIG', DEFAULT_VALUE)` incase env variable not present.

File: app/__init__.py [hunk 2]
```diff
@@ -1,10 +1,72 @@
...
+        {
+            'name': 'WeConnect',
+            'description': 'WeConnect brings businesses and users together, and allows users to review businesses'
+        },
+    ]
+}
+
+swagger = Swagger(app)
+
+# db = SQLAlchemy(app)
```
Comment (collin5) [inline app/__init__.py]: Remove unused code

File: app/__init__.py [hunk 3]
```diff
@@ -1,10 +1,72 @@
...
+db.init_app(app)
+
+# Add Blueprint; how to construct or extend the app
```
Comment (collin5) [inline app/__init__.py]: Comments should be more descriptive

File: app/__init__.py [hunk 4]
```diff
@@ -1,10 +1,72 @@
...
+api_bp = Blueprint('api', __name__)
+api = Api(api_bp, prefix="/api/v2")
+
+
+# Add the resource to the API.
```
Comment (collin5) [inline app/__init__.py]: `Add resources` since they are more than one

File: app/api/v2/resources/auth.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
+# app/api/resources/auth.py
+"""Contains user logic"""
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: This docstring is not needed, we talked about this

File: app/api/v2/resources/auth.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+# use the inspect module because for os.path.abspath(__file__),
+# the __file__ attribute is not always given
+auth_dir = os.path.dirname(os.path.abspath(
+    inspect.getfile(inspect.currentframe())))
+resources_dir = os.path.dirname(auth_dir)
+v2_dir = os.path.dirname(resources_dir)
+api_dir = os.path.dirname(v2_dir)
+app_dir = os.path.dirname(api_dir)
+sys.path.insert(0, app_dir)
+# sys.path.append(os.path.dirname)
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: Remove unused code

File: app/api/v2/resources/auth.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+user_request_parser.add_argument(
+    "password", type=str, required=True,
+    help="Password is required")
+
+
+def string_empty(string_var):
+    """Return true if string is empty"""
+
+    return not isinstance(string_var, str) or string_var in [' ', '']
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: - Docstring and func name is different from what this function does. `string_empty(123)` will return `True` yet passed argument is of type `int`. This is supposed to raise a `TypeError`.

- `string_empty(' ')` will return `False` yet string is empty.

- `or string_var in [' ', '']` can easily be rewritten as `or not string_var.strip()`

File: app/api/v2/resources/auth.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            if authorization:
+                access_token = authorization.split(" ")[1]
+
+            if not access_token:
+                return {"message": "No token provided"}, 401
+
+            if Blacklist.query.filter_by(token=access_token).first():
+                return {"message": "Invalid token provided"}, 401
+
+            try:
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: You can catch all these exceptions in one `try catch` block. 
```
try:
  ...........
except ExceptionA:
 ...........
except ExceptionB:
  ..........
except ExceptionC:
 ..........

and so on

File: app/api/v2/resources/auth.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+def token_required(function):
+    @wraps(function)
+    def decorated_function(*args, **kwargs):
+        access_token = None
+
+        if "Authorization" in request.headers:
+            authorization = request.headers.get("Authorization", None)
+            if authorization:
+                access_token = authorization.split(" ")[1]
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: `KeyError` if `authorization` is on long string i.e if `ABCD`, `authorization.split(" ")[1]` will be out of range.

File: app/api/v2/resources/auth.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+            log_object = Log(
+                "Insert", "Added user: {}".format(username), "users", user_object.id)
+            db.session.add(log_object)
+            db.session.commit()
+
+            # Post create success
+            return make_response(jsonify({"message": "User added"}), 201)
+
+        # Bad request
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: Same here

File: app/api/v2/resources/auth.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            user_object = User(first_name, last_name, username, password)
+            db.session.add(user_object)
+            db.session.commit()
+
+            log_object = Log(
+                "Insert", "Added user: {}".format(username), "users", user_object.id)
+            db.session.add(log_object)
+            db.session.commit()
+
+            # Post create success
```
Comment (collin5) [inline app/api/v2/resources/auth.py]: This comment is not needed

File: config.py [hunk 12]
```diff
@@ -1,22 +1,30 @@
...
+
+
+class Config:
     """
     Common configurations
     """
 
-    SECRET_KEY = 'really secret, is it'
+    SECRET_KEY = os.environ.get('SECRET_KEY', 'some value')
+    SQLALCHEMY_DATABASE_URI = os.environ['DATABASE_URL']
```
Comment (collin5) [inline config.py]: Do the same here as you did above for `SECRET_KEY`.

File: config.py [hunk 13]
```diff
@@ -32,7 +40,6 @@ class TestingConfig(Config):
     Testing configurations
     """
 
-    SECRET_KEY = 'some value'
     TESTING = True
```
Comment (collin5) [inline config.py]: We taked about removing `TESTING` for Production environment

File: app/__init__.py (hunk 1 above)
Comment (Rwothoromo) [reply app/__init__.py]: Noted

File: app/__init__.py (hunk 2 above)
Comment (Rwothoromo) [reply app/__init__.py]: Noted

File: app/__init__.py (hunk 4 above)
Comment (Rwothoromo) [reply app/__init__.py]: Noted

File: app/api/v2/resources/auth.py (hunk 5 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Noted

File: app/api/v2/resources/auth.py (hunk 6 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Noted

File: app/api/v2/resources/auth.py (hunk 7 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Oh yes, true.

I will change it as follows

```
def valid_string(var):
    """Return true if var is string"""
    
    return isinstance(var, str) or not var.strip()
```

File: app/api/v2/resources/auth.py (hunk 8 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Oh, awesome. Let me do that right away.

File: app/api/v2/resources/auth.py (hunk 10 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: I am lost on this one.

File: app/api/v2/resources/auth.py (hunk 11 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Noted

File: config.py (hunk 13 above)
Comment (Rwothoromo) [reply config.py]: I set `TESTING = False` in `class Config` for all environments.
I will add it for production as well.

File: app/api/v2/resources/auth.py (hunk 9 above)
Comment (Rwothoromo) [reply app/api/v2/resources/auth.py]: Noted. I will adjust it to
```
authorization = request.headers.get("Authorization", None)
if authorization:
    authorization_list = authorization.split(" ")
    if len(authorization_list) > 1:
        access_token = authorization_list[1]

    if not access_token:
        return {"message": "No token provided"}, 401
```

File: config.py (hunk 13 above)
Comment (collin5) [reply config.py]: Sorry about that. I thought it was `ProductionConfig`, just realized it's `TestingConfig`

**Pull request merged**